numpy/_core/src/multiarray/descriptor.c

```diff
@@ -1599,6 +1600,13 @@ _convert_from_type(PyObject *obj) {
         return PyArray_DescrFromType(NPY_OBJECT);
     }
     else {
+        PyArray_DTypeMeta *DType = PyArray_DiscoverDTypeFromScalarType(obj);
```

Comment:
Hmmm universal segfault in the test, is this crashing for some type?  I also wonder if we need to call this function also in the first branch, because it should still be fine to subclass from `np.generic`.  Or maybe additionally, if someone registers a DType with a scalar type that subclasses from `np.generic` we insert it into that lookup as well?

Reply:
oops, this is what happens when you make a PR without running the tests first...

Reply:
(Indeed, a scalar type subclassing `np.generic` is exactly what we use for `ml_dtypes` at the moment. In fact, I'd ideally like to subclass `np.floating`, if that turns out not to break things under NumPy 2.0. It confused things pre-NumPy 2.0, because there were assumptions like "there is only one floating point type of a particular width".)

Reply:
If you have a new DType, it may be you get lucky, but there may still be places that use the character code + width, without side-stepping for new ones.
In which case I suppose we may want to fix those if you run into them...